test/functional/p2p_sendrecon.py

```diff
@@ -0,0 +1,166 @@
+    sendrecon_msg.initiator = True
+    sendrecon_msg.responder = False
+    sendrecon_msg.version = 1
+    sendrecon_msg.salt = 2
+    return sendrecon_msg
+
+class SendReconTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [['-txrecon']]
+
+    def run_test(self):
+        # Checks for the node *sending* SENDRECON
+        self.log.info('SENDRECON sent')
+        # Inbound
+        peer = self.nodes[0].add_p2p_connection(SendReconSender(), send_version=True, wait_for_verack=False)
+        peer.wait_until(lambda: peer.sendrecon_msg_received)
+        assert not peer.sendrecon_msg_received.initiator
+        assert peer.sendrecon_msg_received.responder
+        assert_equal(peer.sendrecon_msg_received.version, 1)
+        # Outbound
+        peer = self.nodes[0].add_outbound_p2p_connection(SendReconSender(), p2p_idx=1, connection_type="outbound-full-relay")
+        peer.wait_until(lambda: peer.sendrecon_msg_received)
+        assert peer.sendrecon_msg_received.initiator
+        assert not peer.sendrecon_msg_received.responder
+        assert_equal(peer.sendrecon_msg_received.version, 1)
+
+        self.log.info('SENDRECON should not be sent if block-relay-only')
+        peer = self.nodes[0].add_outbound_p2p_connection(SendReconSender(), p2p_idx=2, connection_type="block-relay-only")
+        peer.wait_for_verack()
+        assert not peer.sendrecon_msg_received
+
+        self.log.info('SENDRECON on pre-WTXID version should not be sent')
+        peer = self.nodes[0].add_p2p_connection(SendReconSender(), send_version=False, wait_for_verack=False)
+        pre_wtxid_version_msg = msg_version()
+        pre_wtxid_version_msg.nVersion = 70015
+        pre_wtxid_version_msg.strSubVer = P2P_SUBVERSION
+        pre_wtxid_version_msg.nServices = P2P_SERVICES
+        pre_wtxid_version_msg.relay = 1
+        peer.send_message(pre_wtxid_version_msg)
+        peer.wait_for_verack()
+        assert not peer.sendrecon_msg_received
+
+        # Checks for the node *receiving* SENDRECON
+        self.log.info('valid SENDRECON')
+        peer = self.nodes[0].add_p2p_connection(
+            SendReconSender(sendrecon_msg_to_send=create_sendrecon_msg()),
+            send_version=True, wait_for_verack=False)
+        peer.wait_until(lambda: peer.version_received)
+        # time.sleep(1)
+        assert peer.is_connected
+        self.log.info('second SENDRECON triggers a disconnect')
+        peer.send_message(create_sendrecon_msg())
+        peer.wait_for_disconnect()
+
+        self.log.info('SENDRECON with initiator=responder=0 triggers a disconnect')
+        sendrecon_no_role = create_sendrecon_msg()
+        sendrecon_no_role.initiator = False
+        sendrecon_no_role.responder = False
+        peer = self.nodes[0].add_p2p_connection(
+            SendReconSender(sendrecon_msg_to_send=sendrecon_no_role),
+            send_version=True, wait_for_verack=False)
```

Comment:
CI fails here, my guess is it's getting disconnected for the sendrecon violation before the test framework checks `peer.is_connected`?